### PR TITLE
`Document` struct with graph sorting and materialisation logic

### PR DESCRIPTION
**Pull request by sandreae**
# Overview

`Document` is a deterministically resolvable data type constructed from a series of causally linked `Operation`. Each operation follows a set schema, and it can be reduced into a final "view" which is the result of applying a list of sorted operations in order.

Internally it builds a graph structure from the operations (which can each contain multiple causal links) and then sorts the graph in a  depth-first topological manner.

### `DocumentBuilder`

The construction and validation of `Document`s occurs in `DocumentBuilder`. This struct is constructed from a list of operations (with some added meta data) and then on call to `build()` checks if the operation collection adheres to certain criteria (exactly one CREATE, operations match one schema) and then builds and returns the `Document`. 

One nice feature of this pattern is that it allows for future extension which could account for documents with different schema (system schema). By defining schema on the document builder, a `Document` with schema specific resolution logic and view could be constructed and returned.

### `Document`

`Document` is an immutable struct returned from `DocumentBuilder` it contains an already resolved document view.

```rust
pub struct Document {
    id: Hash,
    author: Author,
    schema: Hash,
    view: DocumentView,
    meta: DocumentMeta,
}

pub struct DocumentMeta {
    deleted: bool,
    edited: bool,
    operations: Vec<OperationWithMeta>,
    graph_tips: Vec<Hash>,
}
```

### ~`Instance`~ `DocumentView`

A representation of the resolved document state. It contains a k-v map of `OperationValue`s which follow the documents schema definition.

### `Graph`

This is a generic directed acyclic graph which can sort it's nodes using a topological depth first search. The API is (loosely?) based on the `tangle-js` graph implementation but the graph walking algorithm differs.   

_bonus material:_ : `Graph` also derives and returns a list of all branch and graph tips when walking the graph. A "branch tip" means "the tip of any branch which exists in the graph but has been merged" and a "graph tip" means un-merged tips which should be the next operations `previous_operations` (see `Operation` docs if this doesn't make any sense). Combining these two lists gives us a vector clock like record of the exact state of this document. If two document replicas have the same vector-clock then we know they are the same.

## Notes / Changes
- `Document` has now been somewhat promoted. On it's creation it already contains it's resolved `DocumentView` as well as any meta data relating to that specific view.
- `Graph` implements new features for learning about the state of the graph and knowing the current graph tips.

## The future
- Construct documents with custom resolution logic depending on the schema they follow. 

## Last stretch:

- [x] Calculate `current_graph_tip` when resolving `Document` (this will then be used for calculating `next_entry_args`)
  - [x] ~Bonus: also calculate `all_graph_tips` for vectorclock-esque snapshots~ (removed as we aren't sure if we need this)
- [x] handle delete operations
- [x] more meta data on `Document`? 
  - [x] author
  - [x] `is_deleted`, 
  - [x] `is_updated`
  - [x] ~`missing_dependencies`~ for later
  - [x] `graph_tips` 
- [x] Rename `Instance` to `DocumentView`
- [x] Do the docs for Document right!

## 📋 Checklist

- [x] Add tests that cover your changes
- [x] Add this PR to the _Unreleased_ section in `CHANGELOG.md`
- [x] Link this PR to any issues it closes
- [x] New files contain a SPDX license header
- [x] Check if descriptions and terminology match `handbook` content (and visa-versa)

### CODE REVIEW

File: p2panda-rs/src/document/document.rs [hunk 1]
```diff
@@ -0,0 +1,474 @@
...
+        let author = create_operation.public_key();
+
+        // Get the document id from the create message
+        let document_id = create_operation.operation_id();
+
+        // Get the document id from the create message
+        let schema_hash = create_operation.schema();
+
+        // Normally we would get the schema string from the DB by it's hash
+        let schema = Schema::new(&schema_hash, DOCUMENT_SCHEMA)?;
```
Comment (adzialocha) [inline p2panda-rs/src/document/document.rs:202]: This looks a little bit like it needs some rework, probably of the API. I would assume that something gets the `Schema` instance from somewhere.

File: p2panda-rs/src/document/document.rs [hunk 2]
```diff
@@ -0,0 +1,474 @@
...
+
+        // Instantiate graph and operations map
+        let mut graph = IncrementalTopo::new();
+        let mut operations = BTreeMap::new();
+
+        for op in self.operations() {
+            // Validate each operation against the document schema before continuing.
+            // NB. cddl crate not wasm supported yet.
```
Comment (adzialocha) [inline p2panda-rs/src/document/document.rs:210]: This is partly true, the https://crates.io/crates/cddl/0.9.0-beta.0 release has wasm support, but this can be part of another PR.

I would maybe use consistent `@TODO` commenting style, so its easier to find them when searching for todos.

File: p2panda-rs/src/document/document.rs [hunk 3]
```diff
@@ -0,0 +1,474 @@
...
+
+            // Insert operation into map
+            operations.insert(op.operation_id().as_str().to_owned(), op.to_owned());
+            // Add node to graph
+            graph.add_node(op.operation_id().as_str().to_string());
+        }
+
+        // Derive graph dependencies from all operations' previous_operations field. Apply to graph handling
+        // errors.
+        // nb. I had some problems capturing the actual errors from IncrementalTopo crate... needs another
```
Comment (adzialocha) [inline p2panda-rs/src/document/document.rs:222]: Let's look at it together? What errors do you get?

File: p2panda-rs/src/document/document.rs [hunk 4]
```diff
@@ -0,0 +1,474 @@
...
+
+use crate::document::{DocumentBuilderError, DocumentError};
+use crate::hash::Hash;
+use crate::identity::Author;
+use crate::instance::Instance;
+use crate::operation::{AsOperation, OperationWithMeta};
+use crate::schema::{Schema, ValidateOperation};
+use incremental_topo::IncrementalTopo;
+
+use crate::Validate;
```
Comment (adzialocha) [inline p2panda-rs/src/document/document.rs:14]: Small import grouping inconsistency :smile:

File: p2panda-rs/src/document/document.rs [hunk 5]
```diff
@@ -0,0 +1,474 @@
...
+    }
+}
+
+#[cfg(not(target_arch = "wasm32"))]
+impl Validate for Document {
+    type Error = DocumentError;
+
+    fn validate(&self) -> Result<(), Self::Error> {
+        // NB. This validation is quite excessive as it's normally not possible to get to this
+        // point while having broken some of these basic data restraints.
```
Comment (adzialocha) [inline p2panda-rs/src/document/document.rs:107]: Interesting comment. What are the "basic data restraints". Is it worth putting this into an issue?

File: p2panda-rs/src/document/document.rs [hunk 6]
```diff
@@ -0,0 +1,474 @@
...
+    /// The hash id of this document, it is the hash of the entry of this documents root CREATE operation.
+    id: Hash,
+    /// The hash id of the schema operations in this document follow.
+    schema: Schema,
+    /// The author (public key) who published the CREATE message which instantiated this document.
+    author: Author,
+    /// Permissions, derived from KeyGroup relations, which apply to the document as a whole or operation ranges within it.
+    /// TODO: this is a placeholder for now.
+    #[allow(dead_code)]
+    permissions: Option<Vec<Author>>,
```
Comment (adzialocha) [inline p2panda-rs/src/document/document.rs:33]: I would remove everything related to permissions for now, better not mention it at all when its not fleshed out yet :+1:

File: p2panda-rs/src/document/document.rs [hunk 7]
```diff
@@ -0,0 +1,474 @@
...
+        self.operations_iter()
+            .try_for_each(|successor: OperationWithMeta| {
+                if let Some(previous_operations) = successor.previous_operations() {
+                    previous_operations.iter().try_for_each(|previous| {
+                        match graph.add_dependency(
+                            &previous.as_str().to_owned(),
+                            &successor.operation_id().as_str().to_owned(),
+                        ) {
+                            Ok(_) => Ok(()),
+                            Err(_) => Err(DocumentBuilderError::IncrementalTopoDepenedencyError),
```
Comment (adzialocha) [inline p2panda-rs/src/document/document.rs:233]: Typo in: `IncrementalTopoDepenedencyError`

**Review by adzialocha — CHANGES_REQUESTED** [review]
Did a first quick scan, just added some comments and ideas!

**Comment by sandreae** [comment]
Fantastic, thanks!

As a general point on this PR,  `Document` brings together so many of our quite new concepts, I think it's a good point for us to reflect on how we want our API to be and how we see all the parts fitting together. With that in mind, it'd be ideal to make sure everyone has a chance to look over it and give feedback, and also do a session on it together, maybe some co-coding. 

So it's cool if this PR sits here a little while and we get a chance to play around with it I think.

**Comment by adzialocha** [comment]
> As a general point on this PR, Document brings together so many of our quite new concepts, I think it's a good point for us to reflect on how we want our API to be and how we see all the parts fitting together. With that in mind, it'd be ideal to make sure everyone has a chance to look over it and give feedback, and also do a session on it together, maybe some co-coding.

Sounds good, would love some coding and jamming together in the new year! :+1:

File: p2panda-rs/src/document/document.rs (hunk 6 above)
Comment (sandreae) [reply p2panda-rs/src/document/document.rs]: Yep, good shout.

File: p2panda-rs/src/document/document.rs (hunk 5 above)
Comment (sandreae) [reply p2panda-rs/src/document/document.rs:117]: I was thinking of things like a DELETE operation not containing fields, an UPDATE and CREATE containing previous_operations, things like that. Right now we're validating these criteria when building the document, but I'm not sure if it's necessary.

What issue would you want to raise for this? Something like "When does validation occur in the materialisation pipeline"?

File: p2panda-rs/src/document/document.rs (hunk 4 above)
Comment (sandreae) [reply p2panda-rs/src/document/document.rs]: duhhh

File: p2panda-rs/src/document/document.rs (hunk 3 above)
Comment (sandreae) [reply p2panda-rs/src/document/document.rs:224]: It was a borrow checker error, yep, let's check it when we look through all this together.

File: p2panda-rs/src/operation/operation_meta.rs [hunk 8]
```diff
@@ -88,6 +90,20 @@ impl AsOperation for OperationWithMeta {
     }
 }
 
+impl CoreHash for OperationWithMeta {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.operation_id().as_str().hash(state);
```
Comment (adzialocha) [inline p2panda-rs/src/operation/operation_meta.rs:95]: This is returning the hash of the string and not of the operation (aka the id hash) if I see this right

Comment (sandreae) [reply p2panda-rs/src/operation/operation_meta.rs:95]: Yes, it is... This is my first time implementing equality traits, and I went for what seemed like the simplest way to establish equality. If the hashes are equal, the operations are equal. Maybe this is a massive abuse of the trait. In which case I need to apologise to the rust god's 😂

Comment (sandreae) [reply p2panda-rs/src/operation/operation_meta.rs:95]: Oh, but this is the hash trait we're talking about. Kind of same line of thinking though...

Comment (sandreae) [reply p2panda-rs/src/operation/operation_meta.rs:95]: The docs seem to suggest hashing only some key values in a struct is valid: https://doc.rust-lang.org/std/hash/trait.Hash.html

Comment (adzialocha) [reply p2panda-rs/src/operation/operation_meta.rs:95]: The point is that you already have a hash in `instance_id` and you turn this into a string and hash it *again*. I wonder if it would simply be enough the return what is inside of `instance_id` or move the whole hashing logic of `Operation.hash()` there (https://github.com/p2panda/p2panda/issues/32)

Comment (sandreae) [reply p2panda-rs/src/operation/operation_meta.rs:95]: Ah, haha, yeh this is silly. I would like to move the hashing logic to `Hash`, this makes sense to me.

Comment (sandreae) [reply p2panda-rs/src/operation/operation_meta.rs:95]: in progress here: https://github.com/p2panda/p2panda/pull/178

File: p2panda-rs/src/document/document.rs [hunk 9]
```diff
@@ -0,0 +1,428 @@
...
+        if collect_create_operation.len() > 1 {
+            return Err(DocumentBuilderError::MoreThanOneCreateOperation);
+        } else if collect_create_operation.is_empty() {
+            return Err(DocumentBuilderError::NoCreateOperation);
+        }
+
+        let create_operation = collect_create_operation.get(0).unwrap(); // unwrap as we know there is one item
+
+        // Validate the provided schema
+        Schema::new(&create_operation.schema(), DOCUMENT_SCHEMA)?;
```
Comment (cafca) [inline p2panda-rs/src/document/document.rs:164]: This is checking the validity of `DOCUMENT_SCHEMA` right now, there's no check as to whether `DOCUMENT_SCHEMA` belongs to the hash given in the first argument. I've been thinking we should maybe become clearer in what we mean we say 'validates' for example. Maybe change the comment to 

> Validates the provided schema's CDDL definition

File: p2panda-rs/src/document/document.rs (hunk 5 above)
Comment (cafca) [reply p2panda-rs/src/document/document.rs:95]: It feels a bit like schema validation is something that should be done by operations themselves, doesn't it? They contain both the fields and the schema that's supposedly being used. And we eventually won't handle operations whose schema definition we don't have. Maybe `OperationWithMeta` should contain the operation's schema definition and check it in its validation and `Operation` doesn't?

Comment (cafca) [reply p2panda-rs/src/document/document.rs:95]: I moved one of the checks here to operation already :o

File: p2panda-rs/src/document/document.rs (hunk 9 above)
Comment (sandreae) [reply p2panda-rs/src/document/document.rs:156]: Yes, good call!

File: p2panda-rs/src/document/document.rs [hunk 10]
```diff
@@ -0,0 +1,422 @@
...
+impl Validate for Document {
+    type Error = DocumentError;
+
+    fn validate(&self) -> Result<(), Self::Error> {
+        // NB. This validation is quite excessive as it's normally not possible to get to this
+        // point while having broken some of these basic data restraints.
+
+        // Validate each operation in this document.
+        self.iter()?.try_for_each(|op| {
+            // Validate each update operation against the document schema.
```
Comment (cafca) [inline p2panda-rs/src/document/document.rs:99]: The `CREATE` operation should also match its own schema I think

File: p2panda-rs/src/document/document.rs (hunk 3 above)
Comment (sandreae) [reply p2panda-rs/src/document/document.rs]: Not using incremental topo anymore, problem solved.

File: p2panda-rs/src/document/document.rs [hunk 11]
```diff
@@ -0,0 +1,390 @@
...
+        let mut exp_result = BTreeMap::new();
+        exp_result.insert(
+            "name".to_string(),
+            OperationValue::Text("Polar Bear Cafe!!!!!!!!!!".to_string()),
+        );
+
+        // Document should resolve to expected value
+        assert_eq!(instance.raw(), exp_result);
+
+        // Multiple replicas receiving operations in different orders should resolve to same value.
```
Comment (adzialocha) [inline p2panda-rs/src/document/document.rs:298]: Very cool you're testing this as well :+1:

File: p2panda-rs/src/document/document.rs [hunk 12]
```diff
@@ -0,0 +1,390 @@
...
+
+    /// The author of this document.
+    pub fn author(&self) -> Author {
+        self.create_operation.public_key().to_owned()
+    }
+
+    /// The schema for this document.
+    pub fn schema(&self) -> Schema {
+        Schema::new(&self.schema_hash(), DOCUMENT_SCHEMA).unwrap()
+    }
```
Comment (adzialocha) [inline p2panda-rs/src/document/document.rs:53]: Maybe this should be a getter for `schema_hash` instead and we pass in the `schema_definition` in the `OperationBuilder.build()` method instead?

File: p2panda-rs/src/document/document.rs [hunk 13]
```diff
@@ -0,0 +1,390 @@
...
+        self.operations.clone()
+    }
+
+    /// Get an iterator over all operations in this document.
+    pub fn operations_iter(&self) -> std::vec::IntoIter<OperationWithMeta> {
+        self.operations.clone().into_iter()
+    }
+
+    /// Build the document.
+    pub fn build(self) -> Result<Document, DocumentBuilderError> {
```
Comment (adzialocha) [inline p2panda-rs/src/document/document.rs:111]: See comment above! We could pass in the `schema_definition` here (which would then be validated against the given `schema_hash` as @cafca noted.

Comment (sandreae) [reply p2panda-rs/src/document/document.rs:110]: Yes, thank you both! This is much better :+1:

Comment (adzialocha) [reply p2panda-rs/src/document/document.rs:110]: Woho! Then we can finally make `DOCUMENT_SCHEMA` an actual test value :partying_face:

File: p2panda-rs/src/document/document.rs [hunk 14]
```diff
@@ -0,0 +1,406 @@
...
+
+        let document_schema = create_operation.schema();
+
+        // Create instantiate a schema, which validates the provided schema's CDDL definition.
+        let schema = Schema::new(&create_operation.schema(), &self.schema_definition)?;
+
+        // Validate the create message for this document against the document schema.
+        // @TODO: Exclude from wasm target until cddl issues sorted (https://github.com/p2panda/p2panda/issues/99).
+        #[cfg(not(target_arch = "wasm32"))]
+        schema.validate_operation_fields(&create_operation.fields().unwrap())?;
```
Comment (cafca) [inline p2panda-rs/src/document/document.rs:127]: Suggestion to remove this validation as we already know that operations conform to their schema

File: p2panda-rs/src/instance/instance.rs [hunk 15]
```diff
@@ -1,23 +1,25 @@
...
 /// The materialised view of a reduced collection of `Operations` describing a document.
 #[derive(Debug, PartialEq, Default)]
-pub struct Instance(HashMap<String, OperationValue>);
+pub struct Instance(BTreeMap<String, OperationValue>);
 
 impl Instance {
     /// Returns a new `Instance`.
     fn new() -> Self {
-        Self(HashMap::new())
+        Self(BTreeMap::new())
```
Comment (cafca) [inline p2panda-rs/src/instance/instance.rs:18]: I think instances should have a `document: Hash` so we can keep track of which document they are describing no?

**Comment by sandreae** [comment]
Some thoughts after todays jam:
- if DocumentBuilder did already walk the graph and generate an instance as you suggest @cafca 
- then document could already contain an instance, which we would have a getter for (it wouldn't be a result as it already got unwraped in the builder pattern)
- Document could also handle some of the functionality we wanted from schema, like:
  - iterating over the schema fields (we could re-introduce passing the schema to DocumentBuilder)
- all validation would happen in DocumentBuilder, which means Document has no chance to error, this makes it nicer to work with I think.
  
Just very rough thoughts, which are mostly not my own ;-p

File: p2panda-rs/src/instance/instance.rs (hunk 15 above)
Comment (sandreae) [reply p2panda-rs/src/instance/instance.rs:18]: The flow has changed slightly now. `DocumentBuilder` now returns `Documents` which have a `view` field holding an already resolved view of this document. They also have `id` and `schema` fields. :+1: in any case, we have access to document id hash now.

File: p2panda-rs/src/materialiser/graph.rs [hunk 16]
```diff
@@ -244,86 +279,137 @@ impl<'a, T: PartialEq + Clone + std::fmt::Debug> Graph<T> {
...
 #[cfg(test)]
 mod test {
+    use crate::materialiser::graph::GraphData;
+
     use super::Graph;
 
     #[test]
     fn basics() {
         let mut graph = Graph::new();
-        graph.add_node("a", "Wake Up");
```
Comment (adzialocha) [inline p2panda-rs/src/materialiser/graph.rs]: Whaaaaat, I was just reading this this morning and was like "aww, thats such a nice way to understand this test". Why did you take it away? :smile:

Comment (sandreae) [reply p2panda-rs/src/materialiser/graph.rs]: Awwww, sorry, I thought maybe I was just having a silly idea and it didn't actually make sense to others. Also, the simpler node letter naming works better for showing the resolved order. But as you liked it, I'll bring it back for it's own special test ;-p

Comment (sandreae) [reply p2panda-rs/src/materialiser/graph.rs]: It's back ;-p

Comment (adzialocha) [reply p2panda-rs/src/materialiser/graph.rs]: Jaaaa!!

File: p2panda-rs/src/graph/graph.rs [hunk 17]
```diff
@@ -0,0 +1,495 @@
+use std::collections::HashMap;
+use std::fmt::Debug;
+
+use super::GraphError;
+
+/// Directed acyclic casaul graph which can be sorted topologically.
+///
+/// Graph API based on [tangle-graph](https://gitlab.com/tangle-js/tangle-graph).
+#[derive(Debug, PartialEq, Clone)]
+pub struct Graph<T: PartialEq + Clone + Debug>(HashMap<String, Node<T>>);
```
Comment (cafca) [inline p2panda-rs/src/graph/graph.rs:10]: Could be nice to add an example for how to make a `Graph`.

File: p2panda-rs/src/operation/error.rs [hunk 18]
```diff
@@ -10,6 +10,10 @@ pub enum OperationError {
     #[error("operation fields can not be empty")]
     EmptyFields,
 
+    /// Invalid attempt to create a delete operation with fields
+    #[error("DELETE operation must not have fields")]
```
Comment (cafca) [inline p2panda-rs/src/operation/error.rs:14]: "delete" can be lowercase

File: p2panda-rs/src/graph/graph.rs [hunk 19]
```diff
@@ -0,0 +1,495 @@
...
+        self.next.is_empty()
+    }
+
+    /// Returns the key for this node.
+    fn key(&self) -> String {
+        self.key.to_owned()
+    }
+
+    /// Returns a vector of keys for the nodes preceding this node in the graph.
+    fn previous(&self) -> Vec<String> {
```
Comment (cafca) [inline p2panda-rs/src/graph/graph.rs:71]: General Rust thing I was wondering about: wouldn't it be better to generally return references instead of cloning for such getter methods because then we don't need to make a copy in memory whenever we want to read that value?

File: p2panda-rs/src/graph/graph.rs [hunk 20]
```diff
@@ -0,0 +1,495 @@
...
+
+    /// Returns true if this node key is a graph tip.
+    pub fn is_tip_node(&self, key: &str) -> bool {
+        match self.get_node(key) {
+            Some(node) => node.is_tip(),
+            None => false,
+        }
+    }
+
+    // NOT IMPLEMENTED //
```
Comment (cafca) [inline p2panda-rs/src/graph/graph.rs:179]: Do we need this or can we delete the lines?

File: p2panda-rs/src/graph/graph.rs [hunk 21]
```diff
@@ -0,0 +1,495 @@
...
+    pub fn root_node(&self) -> &Node<T> {
+        self.0.values().find(|node| node.is_root()).unwrap()
+    }
+
+    /// Returns the root node key.
+    pub fn root_node_key(&self) -> String {
+        self.0.values().find(|node| node.is_root()).unwrap().key()
+    }
+
+    /// Check if all a nodes dependencies have been visited.
```
Comment (cafca) [inline p2panda-rs/src/graph/graph.rs:192]: Introducing the concept of dependencies here. Do we need it or is it the same as "previous nodes"?

File: p2panda-rs/src/graph/graph.rs [hunk 22]
```diff
@@ -0,0 +1,495 @@
...
+            if sorted_nodes.len() > self.0.len() {
+                return Err(GraphError::CycleDetected);
+            }
+            // Push this node to the sorted stack...
+            sorted_nodes.push(current_node);
+            graph_data.sorted.push(current_node.data());
+            if current_node.is_tip() {
+                graph_data.graph_tips.push(current_node.data())
+            }
+            // println!(
```
Comment (cafca) [inline p2panda-rs/src/graph/graph.rs:249]: Should this be a debug log or be removed?

File: p2panda-rs/src/graph/graph.rs [hunk 23]
```diff
@@ -0,0 +1,495 @@
...
+        };
+        next_nodes.sort_by_key(|node_a| node_a.key());
+        next_nodes.reverse();
+        Some(next_nodes)
+    }
+
+    /// Sorts the graph topologically and returns the sorted
+    pub fn walk_from(&'a self, key: &str) -> Result<GraphData<T>, GraphError> {
+        let root_node = self.get_node(key).unwrap();
+        let mut queue = vec![root_node];
```
Comment (cafca) [inline p2panda-rs/src/graph/graph.rs:229]: It would be very helpful to have a short description of what each of these variables defined at the top mean.

File: p2panda-rs/src/operation/error.rs (hunk 18 above)
Comment (adzialocha) [reply p2panda-rs/src/operation/error.rs:14]: Commented on this in the chat, but I like the uppercase writing: DELETE has a special meaning, it represents the (enum) type of the operation, not a verb.

File: CHANGELOG.md [hunk 24]
```diff
@@ -30,6 +30,7 @@ Not released yet, due to unpublished dependencies: :package: `p2panda-rs`
 - `test_utils` module containing `rstest` fixtures, mock `Node` and `Client` structs, test data helper for `p2panda-js` [#116](https://github.com/p2panda/p2panda/pull/116) `rs`
 - Reconciliation logic /w DAG for materialisation module [#129](https://github.com/p2panda/p2panda/pull/129) `rs`
 - `Instance` which encapsulates the materialised view of a reduced collection of `Operations` [#161](https://github.com/p2panda/p2panda/pull/161) `rs`
+- `Document` for sorting and reducing a graph of `Operations` [#169](https://github.com/p2panda/p2panda/pull/169) `rs`
```
Comment (adzialocha) [inline CHANGELOG.md:33]: This should get a :pancakes:!

File: p2panda-rs/benches/graph.rs [hunk 25]
```diff
@@ -0,0 +1,102 @@
+// Benchmarking tests adapted from [https://github.com/declanvk/incremental-topo/tree/main/benches](https://github.com/declanvk/incremental-topo/tree/main/benches).
```
Comment (adzialocha) [inline p2panda-rs/benches/graph.rs:1]: Do we need license headers for these? I don't know! But maybe we should add them, better to have them than not :-D

File: p2panda-rs/benches/graph_comparison.rs [hunk 26]
```diff
@@ -0,0 +1,171 @@
+// Benchmarking tests adapted from [https://github.com/declanvk/incremental-topo/tree/main/benches](https://github.com/declanvk/incremental-topo/tree/main/benches).
```
Comment (adzialocha) [inline p2panda-rs/benches/graph_comparison.rs:1]: Same here

File: p2panda-rs/src/document/document.rs [hunk 27]
```diff
@@ -0,0 +1,501 @@
...
+pub struct Document {
+    id: Hash,
+    author: Author,
+    schema: Hash,
+    view: DocumentView,
+    meta: DocumentMeta,
+}
+
+#[derive(Debug, Clone, Default)]
+pub struct DocumentMeta {
```
Comment (adzialocha) [inline p2panda-rs/src/document/document.rs:24]: `DocumentMeta`, nice name :+1:

File: p2panda-rs/src/document/document.rs [hunk 28]
```diff
@@ -0,0 +1,501 @@
...
+use crate::graph::Graph;
+use crate::hash::Hash;
+use crate::identity::Author;
+use crate::operation::{AsOperation, OperationWithMeta};
+
+/// A resolvable data type made up of a collection of causally linked operations.
+///
+/// Implements the `Resolve` trait for every different schema we support.
+#[derive(Debug, Clone)]
+pub struct Document {
```
Comment (adzialocha) [inline p2panda-rs/src/document/document.rs:15]: It could be nice to give some doc strings for each field of the `Document` and `DocumentMeta` struct. You know, they show so nicely up when you generate the Rust Docs :sparkles: :wink:

Comment (adzialocha) [reply p2panda-rs/src/document/document.rs:15]: .. OF COURSE with new lines in-between!

File: p2panda-rs/src/document/document.rs [hunk 29]
```diff
@@ -0,0 +1,501 @@
...
+    pub fn schema(&self) -> &Hash {
+        &self.schema
+    }
+
+    /// Get the view of this document.
+    pub fn view(&self) -> &DocumentView {
+        &self.view
+    }
+
+    /// Get the operations contianed in this document.
```
Comment (adzialocha) [inline p2panda-rs/src/document/document.rs:113]: Typo: "contianed"

File: p2panda-rs/src/document/document.rs [hunk 30]
```diff
@@ -0,0 +1,501 @@
...
+    /// Get all operations for this document.
+    pub fn operations(&self) -> Vec<OperationWithMeta> {
+        self.operations.clone()
+    }
+
+    /// Build document. This already resolves the current document view.
+    /// Validate the collection of operations which are contained in this document.
+    /// - there should be exactly one CREATE operation.
+    /// - all operations should follow the same schema.
+
```
Comment (adzialocha) [inline p2panda-rs/src/document/document.rs:155]: Remove newline

File: p2panda-rs/src/document/document.rs [hunk 31]
```diff
@@ -0,0 +1,501 @@
...
+    }
+}
+
+// @TODO: This currently makes sure the wasm tests work as cddl does not have any wasm support
+// (yet). Remove this with: https://github.com/p2panda/p2panda/issues/99
+#[cfg(not(target_arch = "wasm32"))]
+#[cfg(test)]
+mod tests {
+    use rstest::rstest;
+    use std::collections::BTreeMap;
```
Comment (adzialocha) [inline p2panda-rs/src/document/document.rs:212]: Nitpick: Import ordering

File: p2panda-rs/src/document/document.rs [hunk 32]
```diff
@@ -0,0 +1,501 @@
...
+        let expected_op_order = vec![
+            panda_1.to_owned(),
+            panda_2.to_owned(),
+            penguin_1.to_owned(),
+            penguin_2.to_owned(),
+            penguin_3.to_owned(),
+        ];
+        let expected_vector_clock = vec![penguin_entry_3_hash.clone(), panda_entry_2_hash];
+
+        // // Document should resolve to expected value
```
Comment (adzialocha) [inline p2panda-rs/src/document/document.rs:383]: `// //`

File: p2panda-rs/src/document/document_view.rs [hunk 33]
```diff
@@ -0,0 +1,248 @@
...
+    }
+
+    #[rstest]
+    pub fn create_from_schema(#[from(hash)] schema_hash: Hash, create_operation: Operation) {
+        // Instantiate "person" schema from CDDL string
+        let chat_schema_definition = "
+            chat = { (
+                message: { type: \"str\", value: tstr }
+            ) }
+        ";
```
Comment (adzialocha) [inline p2panda-rs/src/document/document_view.rs:230]: I love how everything can become everything!

File: p2panda-rs/src/graph/graph.rs (hunk 19 above)
Comment (sandreae) [reply p2panda-rs/src/graph/graph.rs:71]: Yes, I missed these, thanks!

File: p2panda-rs/src/document/mod.rs [hunk 34]
```diff
@@ -0,0 +1,21 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+//! A Document is a resolvable data type which is made up of a linked graph of operations. Documents MUST have a single root ‘CREATE’
+//! operation. All other operations which mutate the initial data are inserted from this point and, due to the nature of operations,
+//! connect together to form a directed acyclic graph.
```
Comment (adzialocha) [inline p2panda-rs/src/document/mod.rs:5]: I've noticed that Rust Docs usually show the first paragraph of a Docstring in the overview which would be rather long in that case. Just give it a new paragraph after the first sentence, than it should look better :sparkles: 

General note on this text: Its great but also sounds very "speccy", is there a way to reformulate it to address developers "using" this API?

File: p2panda-rs/src/graph/graph.rs [hunk 35]
```diff
@@ -0,0 +1,495 @@
+use std::collections::HashMap;
+use std::fmt::Debug;
+
+use super::GraphError;
+
+/// Directed acyclic casaul graph which can be sorted topologically.
```
Comment (adzialocha) [inline p2panda-rs/src/graph/graph.rs:6]: Typo: "casaul"

File: p2panda-rs/src/graph/graph.rs [hunk 36]
```diff
@@ -0,0 +1,495 @@
...
+    pub fn current_graph_tips(&self) -> Vec<T> {
+        self.graph_tips.clone()
+    }
+    // Returns a list containing all branch tips and the current graph tips.
+    pub fn all_graph_tips(&self) -> Vec<T> {
+        let mut all_graph_tips = self.graph_tips.clone();
+        all_graph_tips.extend(self.merged_branch_tips.clone());
+        all_graph_tips
+    }
+}
```
Comment (adzialocha) [inline p2panda-rs/src/graph/graph.rs]: Style nitpick: Newlines between method definitions

File: p2panda-rs/src/graph/graph.rs (hunk 19 above)
Comment (adzialocha) [reply p2panda-rs/src/graph/graph.rs]: Generally agreeing with your comment, in that case though the contents of a `Vec` need to be owned as far as I know. One could return a reference to the whole `Vec` but that brings ownership problems later on with iterators.

File: p2panda-rs/src/graph/graph.rs (hunk 20 above)
Comment (adzialocha) [reply p2panda-rs/src/graph/graph.rs:172]: Is this because this would be in the tangle API?

File: p2panda-rs/src/graph/graph.rs [hunk 37]
```diff
@@ -0,0 +1,495 @@
...
+
+        if next_nodes.is_empty() {
+            return None;
+        };
+        next_nodes.sort_by_key(|node_a| node_a.key());
+        next_nodes.reverse();
+        Some(next_nodes)
+    }
+
+    /// Sorts the graph topologically and returns the sorted
```
Comment (adzialocha) [inline p2panda-rs/src/graph/graph.rs:219]: This sentence seems to stop suddenly :cat2:

File: p2panda-rs/src/graph/graph.rs [hunk 38]
```diff
@@ -0,0 +1,495 @@
...
+        graph.add_node("b", "Make Coffee");
+        graph.add_node("c", "Drink Coffee");
+        graph.add_node("d", "Stroke Cat");
+        graph.add_node("e", "Look Out The Window");
+        graph.add_node("f", "Start The Day");
+        graph.add_node("g", "Cat Jumps Off Bed");
+        graph.add_node("h", "Cat Meows");
+        graph.add_node("i", "Brain Receives Caffeine");
+        graph.add_node("j", "Brain Starts Engine");
+        graph.add_node("k", "Brain Starts Thinking");
```
Comment (adzialocha) [inline p2panda-rs/src/graph/graph.rs:444]: Awwww, its like pooeetryyy

File: p2panda-rs/src/operation/error.rs (hunk 18 above)
Comment (adzialocha) [reply p2panda-rs/src/operation/error.rs:14]: Which btw, is not very consistent in this PR (hihi)

**Review by adzialocha — CHANGES_REQUESTED** [review]
Incredible work @sandreae! Very nicely written and structured code! Also impressive that on top of all you even updated the mocks.

I only have very pedantic nitpicks, otherwise I can't wait to see this being merged!

**Comment by sandreae** [comment]
I'm noticing all my docs are trash in this PR :rofl:

**Comment by adzialocha** [comment]
> I'm noticing all my docs are trash in this PR rofl

Haha, my review comments are trash!

File: p2panda-rs/src/operation/error.rs (hunk 18 above)
Comment (sandreae) [reply p2panda-rs/src/operation/error.rs:14]: I'm quite attached to the all caps version....

**Review by adzialocha — APPROVED** [review]
